ferminet/pbc/hamiltonian.py

```diff
@@ -0,0 +1,201 @@
+with Fermionic Neural Networks. arXiv preprint arXiv:2202.05183."""
+
+from typing import Callable, Sequence
+
+import chex
+from ferminet import networks
+from ferminet import hamiltonian
+import jax
+import jax.numpy as jnp
+from itertools import product
+
+
+def make_ewald_potential(
+    lattice: jnp.ndarray,
+    atoms: jnp.ndarray,
+    charges: jnp.ndarray,
+    truncation_limit: int = 5,
+    include_heg_background: bool = True
+) -> Callable[[jnp.ndarray, jnp.ndarray], float]:
+  """Creates a function to evaluate infinite Coulomb sum for periodic lattice.
+
+  Args:
+    lattice: Shape (3, 3). Matrix whose columns are the primitive lattice
+      vectors.
+    atoms: Shape (natoms, ndim). Positions of the atoms.
+    charges: Shape (natoms). Nuclear charges of the atoms.
+    truncation_limit: Integer. Half side length of cube of nearest neighbours
+      to primitive cell which are summed over in evaluation of Ewald sum.
+      Must be large enough to achieve convergence for the real and reciprocal
+      space sums.
+    include_heg_background: bool. When True, includes cell-neutralizing
+      background term for homogeneous electron gas.
+
+  Returns:
+    Callable with signature f(ae, ee), where (ae, ee) are atom-electon and
+    electron-electron displacement vectors respectively, which evaluates the
+    Coulomb sum for the periodic lattice via the Ewald method.
+  """
+  rec = 2 * jnp.pi * jnp.linalg.inv(lattice)
+  volume = jnp.abs(jnp.linalg.det(lattice))
+  # the factor gamma tunes the width of the summands in real / reciprocal space
+  # and this value is chosen to optimize the convergence trade-off between the
+  # two sums. See CASINO QMC manual.
+  gamma = (2.8 / volume**(1 / 3))**2
+  ordinals = sorted(range(-truncation_limit, truncation_limit + 1), key=abs)
+  ordinals = jnp.array(list(product(ordinals, repeat=3)))
+  lat_vectors = jnp.einsum('kj,ij->ik', lattice, ordinals)
+  rec_vectors = jnp.einsum('kj,ij->ik', rec, ordinals[1:])
+  rec_vec_square = jnp.einsum('ij,ij->i', rec_vectors, rec_vectors)
+
+  def real_sum(separation: jnp.ndarray):
+    """Real-space Ewald potential between charges seperated by separation.
+    """
+    displacements = jnp.linalg.norm(
+        separation - lat_vectors, axis=-1)  # |r - R|
+    return jnp.sum(
+        jax.scipy.special.erfc(gamma**0.5 * displacements) / displacements)
+
+  def recp_sum(separation: jnp.ndarray):
+    """Reciprocal-space Ewald potential between charges seperated by separation.
+    """
+    return (4 * jnp.pi / volume) * jnp.sum(
+        jnp.exp(1.0j * jnp.dot(rec_vectors, separation)) * \
+        jnp.exp(-rec_vec_square / (4 * gamma)) / rec_vec_square)
+
+  def ewald_sum(separation: jnp.ndarray):
+    """Evaluates combined real and reciprocal space Ewald potential."""
+    return real_sum(separation) + recp_sum(separation) - jnp.pi / (
+        volume * gamma)
+
+  lat_vec_norm = jnp.linalg.norm(lat_vectors[1:], axis=-1)
```

Comment:
Would be more readable if you moved this above `real_sum` so all "top-level" code is grouped together, before the internal functions which close over the variables here.